commands/dial_stdio.go

```diff
@@ -0,0 +1,90 @@
+package commands
+
+import (
+	"io"
+	"os"
+
+	"github.com/docker/buildx/build"
+	"github.com/docker/buildx/builder"
+	"github.com/docker/buildx/util/progress"
+	"github.com/docker/cli/cli/command"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"golang.org/x/sync/errgroup"
+)
+
+type stdioOptions struct {
+	builder string
+}
+
+func runDialStdio(dockerCli command.Cli, opts stdioOptions) error {
+	ctx := appcontext.Context()
+
+	contextPathHash, _ := os.Getwd()
+	b, err := builder.New(dockerCli,
+		builder.WithName(opts.builder),
+		builder.WithContextPathHash(contextPathHash),
+	)
+	if err != nil {
+		return err
+	}
+
+	if err = updateLastActivity(dockerCli, b.NodeGroup); err != nil {
+		return errors.Wrapf(err, "failed to update builder last activity time")
+	}
+	nodes, err := b.LoadNodes(ctx)
+	if err != nil {
+		return err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	printer, err := progress.NewPrinter(ctx, os.Stderr, progressui.AutoMode)
+	if err != nil {
+		return err
+	}
+
+	return progress.Wrap("[internal] Dialing builder: "+b.Name, printer.Write, func(sub progress.SubLogger) error {
+		conn, err := build.Dial(ctx, nodes, printer)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+
+		go func() {
+			<-ctx.Done()
+			conn.Close()
```

Comment:
How does one shut down connection cleanly? If we don't want to add special method then maybe context close should just close the read side of pipe?

Reply:
You mean attempt a half-close like below?

```go

		go func() {
			<-ctx.Done()
			if c, ok := conn.(interface{ CloseRead() error }); ok {
				c.CloseRead()
				return
			}
			conn.Close()
		}()
```

Reply:
No, I meant wrapping `os.Stdin` with `io.Pipe` and then calling `PipeWriter.Close()` on it.

Reply:
Actually, I see you call `conn.Close()` manually. Then yes, that would need to be `CloseWrite()` if conn supports it.

Reply:
Added.